src/util/src/algebra.rs

```diff
@@ -0,0 +1,561 @@
+    ///
+    /// This method return true iff `0 <= self < p` and `self^q % p == 1` where `p` is the group modulus and `q` the group order.
+    pub fn is_valid(&self, group: &Group) -> bool {
+        // It is enough to check the upper bound as self.0 is unsigned.
+        let elem_less_than_p = self.0 < group.p;
+        let elem_has_order_q = self.pow(group.order(), group).0.is_one();
+        elem_less_than_p && elem_has_order_q
+    }
+
+    /// Returns the left padded big-endian encoding of the group element.
+    ///
+    /// The encoding follows Section 5.1.1 in the specs.
+    pub fn to_be_bytes_left_pad(&self, group: &Group) -> Vec<u8> {
+        to_be_bytes_left_pad(&self.0, group.l_p())
+    }
+
+    pub fn to_biguint(&self) -> BigUint {
+        self.0.clone()
+    }
+}
+
+impl Group {
+    /// Constructs a new multiplicative integer group `Z_p^r`.
+    ///
+    /// The arguments are
+    /// - `modulus` - the modulus `p`
+    /// - `cofactor`- the cofactor `r`
+    /// - `generator` - a generator `g`
+    ///
+    /// This function checks that the group is valid according to [`Group::is_valid`].
+    pub fn new(
+        modulus: BigUint,
+        cofactor: BigUint,
+        generator: BigUint,
+        csprng: &mut Csprng,
+    ) -> Option<Self> {
+        let group = Group {
+            p: modulus,
+            r: cofactor,
+            g: generator,
+        };
+        if group.is_valid(csprng) {
+            return Some(group);
+        }
+        None
+    }
+
+    /// Constructs a new group without checking the validity according to [`Group::is_valid`].
+    pub fn new_unchecked(modulus: BigUint, cofactor: BigUint, generator: BigUint) -> Self {
+        Group {
+            p: modulus,
+            r: cofactor,
+            g: generator,
+        }
+    }
+
+    /// This function checks that the given group is valid.
+    ///
+    /// A group is considered valid if:
+    /// - the `modulus` is prime,
+    /// - the `order` is prime, `< modulus-1`, and does not divide the `co-factor`,
+    /// - the `generator` has order `order`,
+    /// - the `co-factor` is non-zero,
+    pub fn is_valid(&self, csprng: &mut Csprng) -> bool {
+        // modulus must be prime
+        if !is_prime(&self.p, csprng) {
+            return false;
+        }
+        // the co-factor must be non-zero
+        if self.r.is_zero() {
+            return false;
+        }
+        let order = (&self.p - BigUint::one()) / &self.r;
+        // the order must be prime, < modulus-1, and must not divide the co-factor
+        if !is_prime(&order, csprng) || self.r.is_one() || (&self.r % &order).is_zero() {
+            return false;
+        }
+        // This ensures that the order of generator is at most `order`
+        if !self.g.modpow(&order, &self.p).is_one() {
+            return false;
+        }
+        // If generator is not one (and `order` is prime) => order of generator is `order`
+        if self.g.is_one() {
```

Comment:
If this check is cheaper than the two previous checks, it could be moved up to fail earlier on cheaper checks. In general, maybe it's worth looking at all checks and arranging them from the cheapest to the most expensive.

Reply:
I have reordered all the checks such that the expensive ones are last.